### PR TITLE
tests, network: Conformance migration test with emulation fix

### PR DESCRIPTION
**Pull request by dteplits**
Conformance migration test with emulation fix

**What this PR does / why we need it**:
Due to https://github.com/kubevirt/kubevirt/issues/6885, network conformance tests fail on environments using emulation.
The failure is caused by VM reboot that is performed as part of migration test flow.
The role of VM reboot in migration test is to rerun DHCP client initialization process and to validate network connection after the migration is completed.
To resolve this issue, VM reboot has been replaced with DHCP client explicit restart which is sufficient to initiate DHCP request to reset VM's IP 

Signed-off-by: Diana Teplits <dteplits@redhat.com>

**Which issue(s) this PR fixes**:
Fixes #
https://github.com/kubevirt/kubevirt/issues/7299

**Special notes for your reviewer**:

**Release note**:
```
NONE
```

### CODE REVIEW

File: tests/network/vmi_networking.go [hunk 1]
```diff
@@ -899,16 +899,37 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
 
-				By("Restarting the vmi")
-				Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
-					&expect.BSnd{S: "sudo reboot\n"},
-					&expect.BExp{R: "reboot: Restarting system"},
-				}, 10)).To(Succeed(), "failed to restart the vmi")
-				tests.WaitUntilVMIReady(vmi, loginMethod)
-				if ipFamily == k8sv1.IPv6Protocol {
-					Expect(configureIpv6(vmi, api.DefaultVMIpv6CIDR)).To(Succeed(), "failed to configure ipv6 on vmi after restart")
+				By("Initiating DHCP server request after migration")
```
Comment (EdDev) [inline tests/network/vmi_networking.go:902]: It is a client request.

File: tests/network/vmi_networking.go [hunk 2]
```diff
@@ -899,16 +899,37 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
+				}, 15)).To(Succeed(), "failed to release dhcp client")
+
+				Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
+					&expect.BSnd{S: "\n"},
+					&expect.BExp{R: console.PromptExpression},
+					&expect.BSnd{S: dhClientUpCmd},
+					&expect.BExp{R: console.PromptExpression},
+					&expect.BSnd{S: tests.EchoLastReturnValue},
+					&expect.BExp{R: console.RetValue("0")},
+				}, 15)).To(Succeed(), "failed to rerun dhcp client")
```
Comment (EdDev) [inline tests/network/vmi_networking.go:929]: Please extract this out to a function, the test reader in this case is more interested in "what" is done, on "whom" and less on the "how".

It also feels awkward that the family is the condition for which DHCP client to use. It smells like the whole test needs to be restructured to make such additions easy to understand later.

File: tests/network/vmi_networking.go [hunk 3]
```diff
@@ -899,16 +899,37 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
+
 				Expect(ping(podIP)).To(Succeed())
+
```
Comment (EdDev) [inline tests/network/vmi_networking.go:932]: nit: Unrelated change.

**Review by EdDev — CHANGES_REQUESTED** [review]
> Due to #6885, network conformance tests fail on environments using emulation.
> The failure is caused by VM reboot that is performed as part of migration test flow.
> The role of VM reboot in migration test is to rerun DHCP client initialization process and to validate network connection after the migration is completed.
> To resolve this issue, VM reboot has been replaced with DHCP client explicit restart which is sufficient to initiate DHCP request to reset VM's IP

I must admit, I did not like this workaround when it was conceived and my opinion has not changed.
I understand the value of passing the conformance tests with the migration tests, but I see a problem with hacking our way out of this by trying to manipulate the scenario.

The issue I see is that accessing the console of the guest and trying to touch the DHCP client is not really fully representing a real scenario. E.g. one may release the lease while the other not, one may clean the cache, one may not.
It is actually unclear what each guest OS does the reboot, so we could argue that the best we can do is to check that whatever it does, it is able to come back functional.

I can try to argue that this specific check that was added, does not really fit a conformance test. Or more accurately, it merits its own test which can just can be excluded from the conformance tests.

Bottom line, I think that a reboot is more realistic and dropping it because of a different bug we have much less production reason to cover is not the best path to take.
Said that, these types of decisions are best to be left to a maintainer, so I will try to just review the code itself.

File: tests/network/vmi_networking.go (hunk 1 above)
Comment (dteplits) [reply tests/network/vmi_networking.go]: done

File: tests/network/vmi_networking.go (hunk 2 above)
Comment (dteplits) [reply tests/network/vmi_networking.go]: simplified the code

File: tests/network/vmi_networking.go [hunk 4]
```diff
@@ -838,6 +838,17 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 				}, timeout, time.Second).Should(Succeed(), fmt.Sprintf("migration should succeed after %d s", timeout))
 			}
 
+			commandDhClient := func(cmd string) error {
```
Comment (EdDev) [inline tests/network/vmi_networking.go:841]: Can this be a regular function?

But maybe you can check if `console.RunCommand` can be used instead.
(you can see usage of it in `PingFromVMConsole` for example)

File: tests/network/vmi_networking.go [hunk 5]
```diff
@@ -857,36 +868,24 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 				}
 			})
 
-			DescribeTable("[Conformance] preserves connectivity", func(ipFamily k8sv1.IPFamily, ports []v1.Port) {
-				libnet.SkipWhenClusterNotSupportIpFamily(virtClient, ipFamily)
+			DescribeTable("[Conformance] preserves connectivity [IPv4]", func(ports []v1.Port) {
```
Comment (EdDev) [inline tests/network/vmi_networking.go:871]: Better not use `[]` as these are used for labeling the tests with keywords (that are later used for filtering).
Or maybe it was your explicit intention?

File: tests/network/vmi_networking.go [hunk 6]
```diff
@@ -857,36 +868,24 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
-				if ipFamily == k8sv1.IPv4Protocol {
-					vmi = masqueradeVMI(ports, "")
-					loginMethod = console.LoginToCirros
-				} else {
-					vmi, err = fedoraMasqueradeVMI(ports, "")
-					Expect(err).ToNot(HaveOccurred(), "Error creating fedora masquerade vmi")
-					loginMethod = console.LoginToFedora
-				}
+				vmi = masqueradeVMI(ports, "")
+				loginMethod = console.LoginToCirros
```
Comment (EdDev) [inline tests/network/vmi_networking.go:879]: No need to this variable anymore, you can inline it.

File: tests/network/vmi_networking.go [hunk 7]
```diff
@@ -899,21 +898,56 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
+			It("[Conformance] should preserve connectivity [IPv6]", func() {
+				libnet.SkipWhenClusterNotSupportIpFamily(virtClient, k8sv1.IPv6Protocol)
+
+				var err error
+				var loginMethod console.LoginToFunction
+
+				By("Create VMI")
+				vmi, err = fedoraMasqueradeVMI([]v1.Port{}, "")
+				Expect(err).ToNot(HaveOccurred())
+				loginMethod = console.LoginToFedora
```
Comment (EdDev) [inline tests/network/vmi_networking.go:920]: Same, you can inline now.

File: tests/network/vmi_networking.go [hunk 8]
```diff
@@ -899,21 +898,56 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
-				Entry("IPv4 with explicit ports used by live migration", k8sv1.IPv4Protocol, portsUsedByLiveMigration()),
-				Entry("IPv6", k8sv1.IPv6Protocol, []v1.Port{}),
+				Entry("IPv4", []v1.Port{}),
+				Entry("IPv4 with explicit ports used by live migration", portsUsedByLiveMigration()),
 			)
+
+			It("[Conformance] should preserve connectivity [IPv6]", func() {
+				libnet.SkipWhenClusterNotSupportIpFamily(virtClient, k8sv1.IPv6Protocol)
+
+				var err error
```
Comment (EdDev) [inline tests/network/vmi_networking.go:914]: Not needed anymore.

File: tests/network/vmi_networking.go [hunk 9]
```diff
@@ -899,21 +898,56 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
+				loginMethod = console.LoginToFedora
+
+				vmi, err = virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(vmi)
+				Expect(err).ToNot(HaveOccurred())
+				vmi = tests.WaitUntilVMIReady(vmi, loginMethod)
+				virtHandlerPod, err := getVirtHandlerPod()
+				Expect(err).ToNot(HaveOccurred())
+
+				err = configureIpv6(vmi, api.DefaultVMIpv6CIDR)
+				Expect(err).ToNot(HaveOccurred())
```
Comment (EdDev) [inline tests/network/vmi_networking.go:929]: Can be oneliner with the use of `Success()`.

File: tests/network/vmi_networking.go [hunk 10]
```diff
@@ -592,6 +592,7 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 	})
 
 	Context("VirtualMachineInstance with masquerade binding mechanism", func() {
+
```
Comment (EdDev) [inline tests/network/vmi_networking.go:595]: Not a related change.

**Review by EdDev — CHANGES_REQUESTED** [review]
Very nice, thank you!

There is some room to extract some common part to functions, but we can do that in a following change. The tests look cleaner this way.

File: tests/network/vmi_networking.go (hunk 5 above)
Comment (dteplits) [reply tests/network/vmi_networking.go:871]: Yes, that was the intention

Comment (EdDev) [reply tests/network/vmi_networking.go:859]: OK, so who is using this label now?

Comment (EdDev) [reply tests/network/vmi_networking.go:859]: [This](https://github.com/kubevirt/kubevirt/blob/f5973335f4fb4f504c0513a42cb54841528bdb4d/automation/test.sh#L365) is how a label is used. As you are not using it for that purpose, I think you should drop the parenthesis.

File: tests/network/vmi_networking.go [hunk 11]
```diff
@@ -899,21 +884,54 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
+				Expect(console.RunCommand(vmi, "sudo cirros-dhcpc down eth0\n", time.Second*time.Duration(15))).To(Succeed(), "failed to release dhcp client")
+				Expect(console.RunCommand(vmi, "sudo cirros-dhcpc up eth0\n", time.Second*time.Duration(15))).To(Succeed(), "failed to run dhcp client")
+
 				Expect(ping(podIP)).To(Succeed())
 			},
-				Entry("IPv4", k8sv1.IPv4Protocol, []v1.Port{}),
-				Entry("IPv4 with explicit ports used by live migration", k8sv1.IPv4Protocol, portsUsedByLiveMigration()),
-				Entry("IPv6", k8sv1.IPv6Protocol, []v1.Port{}),
+				Entry("IPv4", []v1.Port{}),
+				Entry("IPv4 with explicit ports used by live migration", portsUsedByLiveMigration()),
```
Comment (EdDev) [inline tests/network/vmi_networking.go:895]: The `IPv4` appears now in the table description.

**Review by EdDev — APPROVED** [review]
Thanks!

There is another nit, see inline.

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by AlonaKaplan** [comment]
/approve
Thanks!
